### ISSUE TITLE
schemas: consider renaming 'union' to 'variant'

### ISSUE THREAD

**Issue by warpfork** [ipld/ipld#153]
((This is a notes-gathering issue first; it does not necessarily represent immediate intent to make changes.))

I'd like to propose we consider renaming "unions" in IPLD Schemas to "variants".

The reason I'd like to consider this is that it might improve clarity, reduce confusion, and ever-so-slightly smooth adoption (by reducing the incidence rate of pushback from people who might get an incorrect first impression from this term).

This is going to be a long issue, and inherently difficult to read, because it's discussing aligning {words} to {meaning}, and we have to use words to do it.  It's a bit a bootstrapping problem.  Prepare your brain to operate in a slightly mushy mode: thinking should go from words to concepts, then try to forget all the words, and then finally try to reattach words to the concepts and see if they make fresh sense.

## history: how did we get this name in the first place?

Well, it was me, first of all.  So it's worth noting: I've updated my position.  (In other words: This is me repainting my _own_ bikeshed, even with full knowledge of why I picked the old color... which I think gives it a little more credence than a random painter coming by and complaining about the bikeshed color, heh.)

I saw the word "union" used for this concept in a bunch of places and thought I'd roll with it too.

## reasons for concern

After using this term for a while, I'm less thrilled with it, because I notice that it provokes confusion more often than I'd like.  It seems to need a lot of disambiguation work, and that puts a lot of burden on our docs and slows down communication.

- "unions" is a loaded term because historically it means multiple things.  It can be "tagged unions" or "untagged unions".  (Or "discriminated unions" vs "undiscriminated unions".  Many other similar synonyms have been used, too.)
  - ... and it's worth noting that these are fairly _completely different_ things.  Tagged unions are sum types: they explicitly say _which_ member of union the data is, so they're generally safe and clear to work with.  Untagged unions a totally different beast: they're just a heap of data that sorta says "good luck deciding how to make sense of me"; they're not sum types, and they're generally not at all safe nor clear to work with!

- It turns out not everyone is comfortable with assuming "unions" means "tagged unions".  I've had more than a few people make a nasty face at me when I describe IPLD schemas as having unions -- when I ask why, it's invariably because they think of untagged unions (often, because they experienced them in C).  After we discuss that ("no, no, gosh, no, I mean tagged unions, of course!  We're not mad"), it's fine, we get back on the same page, and the other party is still interested rather than rejecting.  But it would be awfully nice not to have this potential bounce!

- If we try to fix this by saying "tagged unions" or "discriminated unions" everywhere... it's just plain a mouthful.  It's a significant barrier.
  - Beyond just being a mouthful, it suggests we'll talk often about some kind of union that isn't those -- and we don't: we don't often spend a ton of time describing _untagged_ unions, because those aren't something IPLD Schemas have any explicit support for.  In other words, having an "X Y" term where there's no "!X Y" term really seems strange.

- If we try to fix this by just boldly holding course, we not only have to keep explaining it, we also are sorta squatting on another legitimate term.  By example: someone recently described a protocol design choice to me as (paraphrasing) "this struct is sort of a union of all the fields in these other two types, just lots of optional fields"... and I nodded, and instantly knew what that person meant.  (They're describing an untagged union, more or less.)  But this is not the way we've used the word "union" when we used it as a name for a typekind IPLD schemas, so this could be very confusing to a listener!  Even though we don't have a feature in IPLD schemas for what this person was describing (on purpose: we discourage this pattern), it _does_ describe something that sometimes occurs in protocol design, so it ought to be describable; and for us to squat a word that's likely used to describe that phenomenon is not good.  And worse, having the word for the alternative that we _recommend_ be overlapping and squatting on a word that's descriptive of what we _don't recommend_ is just awful!

## thinking through a new name

If we picked the name for this concept and feature again now, what would we pick?

### what do other systems call it?

The term in type theory literature is "sum types", hands down.  (And I really like this term: once you grok "cardinality" of data structures, and someone explains that "structs are 'product types'", everything goes *click* and makes a *lot* of sense.  (Including _why_ you should use sum types whenever possible rather than heaping semi-overlapping things together in a struct-i-might-call-an-untagged-union when doing protocol design: to minimize the number of cases you have to test!  But I digress.))

Some type theory literature also may call it "coproducts".  (This makes sense if you've steeped your brain in a lot of type theory literature, but I'd say it's pretty inscruitable at first glance, otherwise.)

OCaml calls them variants: https://v1.realworldocaml.org/v1/en/html/variants.html   A lot of other ML-family languages use this term too.

I think python's gradual typing systems probably calls this a "metaclass": https://mypy.readthedocs.io/en/latest/metaclasses.html ... although I think that may be lumping in a bunch of other things, too, and so is probably not a great direct comparison.  Python also seems to be calling some specific features unions as of [PEP-604](https://www.python.org/dev/peps/pep-0604/).

Rust calls them enums.  (This one really bugs me -- I think an enum and a sum type should be treated as considerably different things.  An enum should be countable.  A sum type often isn't, because it can contain recursive complexity, including product types (ok, still countable) but also things like strings (now it's infinite).  Enums are a degenerate case of sum types; but the opposite is not true.  Lumping sum types into "enums" is about as close to Objectively Wrong as I think one can get on this matter.  *deep breaths* Moving on!)

JSON Schemas have a [oneOf](https://json-schema.org/understanding-json-schema/reference/combining.html#oneof)... but this seems to be more like an untagged union.

OpenAPI 3 lets you use [oneOf](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/)... either with or without a [discriminator](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md#discriminatorObject), so that's fun: they've got both tagged and untagged modes of operation here.  (Sidenote: their discriminator support only appears to be what we'd call "inline unions" in IPLD schema terminology today -- the worst kind, because they're the nastiest to parse efficiently.  Ah well.)

Protobuf 3 introduced [oneof](https://developers.google.com/protocol-buffers/docs/proto3#oneof), which is doing tagged unions.  (Amusingly, note this is lowercase, whereas in OpenAPI and JSON Schemas it was camelCase.)

TypeScript seems to call them unions, or possibly [discriminated unions](https://basarat.gitbook.io/typescript/type-system/discriminated-unions).  (I'm not actually sure I've found "the" definitive reference for what these are called in TypeScript.  The syntax is `x | y` -- so they've avoided using a keyword in the source form.)

GraphQL has a concept of [union](https://graphql.org/learn/schema/#union-types).  (I'm not sure exactly how to file these, because GraphQL has a dichotomy between the types vs how you actually access them with the query language, and that makes things kinda heady.  I think we'd call these tagged unions, though -- the "`__typename`" field in the query is pretty clearly a discriminator tag.)

Since we've mentioned a lot of things calling things "union", though, it's important to mention at least one that uses that word, but uses it *very* differently: C has unions.  But they're _untagged_: they're _not sum types_.  There's a [stack overflow question](https://stackoverflow.com/questions/42844423/writing-a-safe-tagged-union-in-c) that's a great illustration of how in C, you have to put a "union" next to an enum, together in a struct, to build your own sum types.  (This is a big deal, because a *lot* of people have heard of it -- and have been told to *fear the word "union"* because of how easy it is to do this wrong.  I can't speak to everyone's education: but personal anecdote: when I was learning to program, someone told me to fear unions a long time before anyone really taught me clearly what unions _were_ (by _any_ definition of the term).  I'm afraid that might not be uncommon.  Seealso the anecdata earlier about my concern with the frequency that I see a nasty face when using the word "union" without backing up and explaining things.)

Wikipedia has an opinion: [Variant type](https://en.wikipedia.org/w/index.php?title=Variant_type&redirect=no) is a redirect page to "tagged union".  (On the other hand, it also has pages for both "union type" and "tagged union", and they seem rather... fuzzily disambiguated to me.  The [Syntax and Examples section of the Union Type page](https://en.wikipedia.org/wiki/Union_type#Syntax_and_example) combines tagged and untagged unions willy-nilly -- and maximally confusingly, describes even the untagged ones, in the C example, as "tagged": then calls the ones in other languages that _are_ tagged "implicitly" tagged.  I got dizzy reading this, even though I'm fairly sure I already know what's going on with this topic!  I can't imagine how any newcomer to that article would walk away without confusion.  So, as of the current date, I'm not sure Wikipedia's content of this matter ranks very highly in my estimation.)

### so what can we draw from the above?

First of all, no one agrees on anything.

There are lots of options that have prior use.

It seems worth noticing that only two options above that I was able to comment on _without_ a bunch of negative or confusing parantheticals attached to it were "variant" and "sum type".  Both are untainted: I've never seen them used incorrectly or confusingly.

"union" has some sizable traction.  But "union" is also titanically burdened with confusion.

There does seem to be _some_ motion of newer languages to tend to say "union" and mean "tagged union".  (So our initial stance in IPLD Schemas isn't without company.)  But I'm not sure how fast or definite we can say that motion is.  The fact I had to phrase the first sentence of this paragraph with as many hedge words as "seem", "some", and "tend", plus a "say"-vs-"mean"... is not great.

A lot of newer languages have used symbols rather than keywords, which makes the issue less pressing -- e.g. somthing like `x | y` instead of `union{x, y}`.  (Unfortunately, I don't think any kind of dodge like that is possible for us in IPLD schemas.  Not only did we already put a keyword in our DSL syntax, there's no way to avoid *frequently* referring to the typekind keyword in the docs due to how often representation strategies have to talk about it.)

"oneof" is starting to show up in a couple places.  But it has inconsistent capitalization (of course it does), and it's also already been used to describe both tagged and untagged scenarios, so it may be on its way to the same confusion territory as "union" is stuck in.

### narrow it down and pick already

Okay.

We want something that doesn't confuse.  Let's say that scratches "union" out of the running.

We can also scratch out "enum" (Rust is Objectively Wrong about this).  And I don't really know if "metaclass" even belonged in the list above in the first place, so scratch that too.  "coproduct" just isn't appealing; let's scratch that.

Of all the terms remaining: I would like something that's *one word*, because we use this as a syntactic keyword, and it should also be effective in search engine queries.  The remaining terms are "sum type", "variant type", and "oneOf".

"sumtype" is a bit of a stutter when considering how it would be used in syntax (we'd end up having to say "a sumtype typekind" sometimes -- ew.)  "sum" could be used alone, but doesn't make much sense (too likely to connote maths).

"oneOf"... hm, okay, we'll ignore the fact it's always been two words slammed together; it seems acceptable as a keyword.  It's awfully unfortunate that of the two other systems I've seen using it, they already picked different capitalization styles, though.  And the fact some systems that use this term use it to describe untagged behaviors makes me nervous; that's the confusion we're trying to get _away_ from.

"variant" still stands by itself pretty okay.  It's acceptable as a keyword.  I've never seen any confusion about it.  It's never been tainted with conflictory meanings.  I have no complaints.

So.  "variant".

## would this be a big change?

Yes, I'm afraid it probably is.  The amount of both code and prose which refers to the current name is substantial.  Even if the change is literally "sed-level complexity", it would still have to be applied in _many_ places at this point, and those places are widespread.

I think we could still pull it off, but just barely.

One thing to note as _not_ having inertia: I think there are not any places where an IPLD Schema, in either DMT or DSL form, is yet "on a chain" anywhere.  They're still all showing up in dev tooling and git repos, where there are well-understood ways of changing things, and fewer expectations about eternality.  If this were not the case, the inertia would be even higher.

We'd still have effectively breaking changes for:
-  anything using schema related code (e.g. [this](https://pkg.go.dev/github.com/ipld/go-ipld-prime/schema#UnionRepresentation) would be renamed, among various other things);
- anything using codegen (it would both have to change the symbol names for things directly talking about variants, and also presumably may change some other naming conventions in its own type emission);
- and we'd need to apply the sed `s/union/variant/g` to schemas (both DSL and DMT form) in the wild.

The change complexity itself is low.  But the coordination required would be noticeable.

We would also have existing discussions (issues, etc) referring to unions which would not be easy to update.  I don't know how heavily this is worth weighting, though: I tend to assume that the amount of discussions that haven't happened yet is much greater than the amount that have already happened, and therefore easing the future ones is often more valuable than worrying about the discoverability of keywords in older ones.  We would certainly need to write a linkable heading in the docs website about the name change.  But that may be sufficient.

## tl;dr

I think we'd have an easier time explaining things if we used the term "variant".  If I had a time machine: I'd definitely have started using that term in the beginning.  We don't have a time machine.  But maybe we should still consider if it's possible to change to it.

**Comment by warpfork** [comment]
The Bytecode Alliance WIT project has concepts of both "[union](https://github.com/bytecodealliance/wit-bindgen/blob/d17f6734422a66ef945710c710e468cc1b0fbc56/WIT.md#item-union-variant-but-with-no-case-names)" and "[variant](https://github.com/bytecodealliance/wit-bindgen/blob/d17f6734422a66ef945710c710e468cc1b0fbc56/WIT.md#item-variant-one-of-a-set-of-types)" -- and also seems to align on using the word "variant" as I describe above, so I'd submit this as more reason to favor this renaming.

**Comment by Gozala** [comment]
Alternate could be to embrace `variant` and keep `union` as a legacy to avoid breaking things.
